reset.c

```diff
@@ -0,0 +1,593 @@
+ *  @param args - string containing dump of command line invocation
+returns 1 on failure 0 on success
+*/
+int reset(samFile *infile, samFile *outfile, conf_data *config, char *args)
+{
+    sam_hdr_t *in_samhdr = NULL, *out_samhdr = NULL;
+
+    if (!infile || !outfile) {
+        fprintf(stderr, "Invalid parameters in reset!\n");
+        return EXIT_FAILURE;
+    }
+
+    //read input header
+    in_samhdr = sam_hdr_read(infile);
+    if (!in_samhdr)
+    {
+        fprintf(stderr, "Failed to read header from file!\n");
+        return EXIT_FAILURE;
+    }
+    //create output header
+    if (!(out_samhdr = sam_hdr_init()))
+    {
+        fprintf(stderr, "Failed to create output header!\n");
+        sam_hdr_destroy(in_samhdr);
+        return EXIT_FAILURE;
+    }
+    //TODO SO/GO status not maintained, do they hold as secondary n supplementary are removed?
+    //add version to output header
+    if  (-1 == sam_hdr_add_line(out_samhdr,"HD", "VN", SAM_FORMAT_VERSION, NULL)) {
+        fprintf(stderr, "Failed to set header data!\n");
+        sam_hdr_destroy(in_samhdr);
+        sam_hdr_destroy(out_samhdr);
+        return EXIT_FAILURE;
+    }
+    //add RG / PG lines if configured
+    if ((config->keepRGs && getRGlines(in_samhdr, out_samhdr)) ||
+            getPGlines(in_samhdr, out_samhdr, config, args)) {
+        sam_hdr_destroy(in_samhdr);
+        sam_hdr_destroy(out_samhdr);
+        return EXIT_FAILURE;
+    }
+
+    //write output header
+    if (sam_hdr_write(outfile, out_samhdr)) {
+        print_error_errno("reset", "Output header write failed (%d)!\n", errno);
+        sam_hdr_destroy(in_samhdr);
+        sam_hdr_destroy(out_samhdr);
+        return EXIT_FAILURE;
+    }
+
+    int ret_r = 0, ret_w = 0;
+    bam1_t *bamdata = bam_init1();      //input bam
+    bam1_t *outdata = bam_init1();      //output bam
+    if (!bamdata || !outdata)
+    {
+        fprintf(stderr, "Failed to allocate data memory!\n");
+        sam_hdr_destroy(in_samhdr);
+        sam_hdr_destroy(out_samhdr);
+        if (!bamdata)
+            bam_destroy1(bamdata);
+        if (!outdata)
+            bam_destroy1(outdata);
+        return EXIT_FAILURE;
+    }
+
+    errno = 0;
+    int cnt = 0, i = 0;
+    kstring_t querydata = KS_INITIALIZE, qualdata = KS_INITIALIZE;
+    char *sp = NULL, *qp = NULL;
+    uint8_t *bamquery = NULL, *bamqual = NULL;
+    //get bam data, make updates and dump to output
+    while (0 <= (ret_r = sam_read1(infile, in_samhdr, bamdata)))
+    {
+        sp = NULL; qp = NULL;
+        bamquery = NULL; bamqual = NULL;
+
+        // read data
+        cnt++;
+        if (bamdata->core.flag & BAM_FSECONDARY || bamdata->core.flag & BAM_FSUPPLEMENTARY) {
+            continue;
+        }
+
+        //update flags
+        uint16_t flags = bamdata->core.flag & ~BAM_FPROPER_PAIR;    //reset pair info
+        flags |= BAM_FUNMAP | BAM_FMUNMAP;                          //mark as unmapped
```

Comment:
This shouldn't set `BAM_FMUNMAP` if the data was originally not paired (eg single-ended ONT data).

Reply:
Corrected.